### ISSUE TITLE
"dot1x_enable" is handled incorrectly in switchport_feature_template?

### ISSUE THREAD

**Issue by cmohorea** [CiscoDevNet/terraform-provider-sdwan#219]
#1: when creating a resource, no matter if "dot1x_enable" set to true or false, it is implemented/shows in vManage as "enabled"
#2: much worse, when attaching to the device, it works the 1st time, and then next runs, TF tries to set "dot1x_enable" values again (with "+", i.e. create it again), and fails, so generally cannot use Terraform for the deployment anymore. Notably, the state in the "tfstate" file is correct

sample resource:
```
resource "sdwan_switchport_feature_template" "SITE_SWITCHPORT_IR1101_FE00" {
  name = "SITE_SWITCHPORT_IR1101_FE00"
  description = "IR1101 4 port FastEthernet switch module"
  device_types = ["vedge-IR-1101"]
  module_type = "4"
  slot = 0
  sub_slot = 0
  interfaces =   [
    {
      name = "FastEthernet0/0/1",
      shutdown_variable = "FE0_0_1_SHUT",
      switchport_mode = "access",
      switchport_access_vlan_variable = "FE0_0_1_ACCESS_VLAN",
      dot1x_enable = false,
    },
    {
      name = "FastEthernet0/0/2",
      shutdown_variable = "FE0_0_2_SHUT",
      switchport_mode = "access",
      switchport_access_vlan_variable = "FE0_0_2_ACCESS_VLAN",
      dot1x_enable = true,
    },
    {
      name = "FastEthernet0/0/3",
      shutdown_variable = "FE0_0_3_SHUT",
      switchport_mode = "access",
      switchport_access_vlan_variable = "FE0_0_3_ACCESS_VLAN",
      dot1x_enable = false,
    },
    {
      name = "FastEthernet0/0/4",
      shutdown_variable = "FE0_0_4_SHUT",
      switchport_mode = "trunk",
      switchport_trunk_allowed_vlans_variable = "FE0_0_4_TRUNK_ALLOW_VLAN",
      switchport_trunk_native_vlan_variable = "FE0_0_4_TRUNK_NATIVE",
      dot1x_enable = false,
    }
  ]
}

```

Implementation:
1st run: all OK, resource created and attached:
```
  # sdwan_switchport_feature_template.SITE_SWITCHPORT_IR1101_FE00 will be created
  + resource "sdwan_switchport_feature_template" "SITE_SWITCHPORT_IR1101_FE00" {
      + description   = "IR1101 4 port FastEthernet switch module"
      + device_types  = [
          + "vedge-IR-1101",
        ]
      + id            = (known after apply)
      + interfaces    = [
          + {
              + dot1x_enable                    = false
              + name                            = "FastEthernet0/0/1"
              + shutdown_variable               = "FE0_0_1_SHUT"
              + switchport_access_vlan_variable = "FE0_0_1_ACCESS_VLAN"
              + switchport_mode                 = "access"
            },
          + {
              + dot1x_enable                    = true
              + name                            = "FastEthernet0/0/2"
              + shutdown_variable               = "FE0_0_2_SHUT"
              + switchport_access_vlan_variable = "FE0_0_2_ACCESS_VLAN"
              + switchport_mode                 = "access"
            },
          + {
              + dot1x_enable                    = false
              + name                            = "FastEthernet0/0/3"
              + shutdown_variable               = "FE0_0_3_SHUT"
              + switchport_access_vlan_variable = "FE0_0_3_ACCESS_VLAN"
              + switchport_mode                 = "access"
            },
          + {
              + dot1x_enable                            = false
              + name                                    = "FastEthernet0/0/4"
              + shutdown_variable                       = "FE0_0_4_SHUT"
              + switchport_mode                         = "trunk"
              + switchport_trunk_allowed_vlans_variable = "FE0_0_4_TRUNK_ALLOW_VLAN"
              + switchport_trunk_native_vlan_variable   = "FE0_0_4_TRUNK_NATIVE"
            },
        ]
      + module_type   = "4"
      + name          = "SITE_SWITCHPORT_IR1101_FE00"
      + slot          = 0
      + sub_slot      = 0
      + template_type = (known after apply)
      + version       = (known after apply)
    }
.....................
Apply complete! Resources: 2 added, 3 changed, 0 destroyed.
```

Re-doing "apply" immediately after:
```
  # sdwan_switchport_feature_template.SITE_SWITCHPORT_IR1101_FE00 will be updated in-place
  ~ resource "sdwan_switchport_feature_template" "SITE_SWITCHPORT_IR1101_FE00" {
        id            = "a3175c47-f695-4dcd-bd38-0c32ac932e01"
      ~ interfaces    = [
          ~ {
              + dot1x_enable                    = false
                name                            = "FastEthernet0/0/1"
                # (3 unchanged attributes hidden)
            },
          ~ {
              + dot1x_enable                    = true
                name                            = "FastEthernet0/0/2"
                # (3 unchanged attributes hidden)
            },
          ~ {
              + dot1x_enable                    = false
                name                            = "FastEthernet0/0/3"
                # (3 unchanged attributes hidden)
            },
          ~ {
              + dot1x_enable                            = false
                name                                    = "FastEthernet0/0/4"
                # (4 unchanged attributes hidden)
            },
        ]
        name          = "SITE_SWITCHPORT_IR1101_FE00"
      ~ version       = 0 -> (known after apply)
        # (6 unchanged attributes hidden)
    }
......................................
╷
│ Error: Client Error
│
│   with sdwan_attach_feature_device_template.LAB1_SITE4_WAYSIDE_DIRECT-1,
│   on cnrail-lab1-branch-pod4.tf line 135, in resource "sdwan_attach_feature_device_template" "LAB1_SITE4_WAYSIDE_DIRECT-1":
│  135: resource "sdwan_attach_feature_device_template" "LAB1_SITE4_WAYSIDE_DIRECT-1" {
│
│ Failed to configure object (POST), got error: HTTP Request failed: StatusCode 400,
│ {"error":{"type":"error","message":"Template edit request has expired.","details":"Please try
│ again.","code":"MTEMP0015"}}
╵
╷
│ Error: Client Error
│
│   with sdwan_attach_feature_device_template.LAB1_SITE4_WAYSIDE_DIRECT,
│   on cnrail-lab1-branch-pod4.tf line 183, in resource "sdwan_attach_feature_device_template" "LAB1_SITE4_WAYSIDE_DIRECT":
│  183: resource "sdwan_attach_feature_device_template" "LAB1_SITE4_WAYSIDE_DIRECT" {
│
│ Failed to configure object (POST), got error: HTTP Request failed: StatusCode 400,
│ {"error":{"type":"error","message":"Template edit request has expired.","details":"Please try
│ again.","code":"MTEMP0015"}}
╵
```

**Comment by cmohorea** [comment]
FYI, workaround to #2: go to vManage and set "802.1X" to "Off" manually (to match TF setup), then TF won't try to make changes

**Comment by danischm** [comment]
Thanks for opening the issue! There are some boolean values where `false` is also interpreted as "enabled". As a workaround can you try setting `dot1x_enable` to null or not set it at all and see if it makes a difference?

**Comment by cmohorea** [comment]
all 3 states (true, false, null) result in "On" set in vManage.
I am not sure if "null" would be acceptable in general as there's no "default" option for this in GUI:
![image](https://github.com/CiscoDevNet/terraform-provider-sdwan/assets/13035952/fd5ab2e8-dd49-45b1-8d29-9172c6cd490e)

**Comment by danischm** [comment]
This should disable it with "null": https://github.com/CiscoDevNet/terraform-provider-sdwan/commit/37721cc6199c424b7c49652fa7047e3f26c31968

**Comment by cmohorea** [comment]
I hope it also will be disabled with "false".  
Null corresponds to "default" value, and there's no "default" option when configuring in vManage. 
What would be the value when you do "import" of the existing template with this value set to "Off" ?

Will be happy to test when released though!

**Comment by danischm** [comment]
Currently a boolean value of "false" - in some instances - is not correctly translated according to the schema by vManage. At this point I am not sure this is something I want to fix in the provider, therefore the current workaround is to use "null". "Off" should be represented by a non-existing value in the payload (you should be able to confirm this by looking at the payload generated when creating the template via the GUI), which should then be translated to "null" again.

**Comment by cmohorea** [comment]
I see same behavior in 0.3.8:
1) Create a new feature template with 3 interfaces set with `dot1x_enable` set to  false/true/null
2) terraform apply => tfstate reflects state as per config, while in vManage all 3 are shown as Dot1X = On
3) terraform apply again, now tfstate and real state are different:
```
$ terraform -v
Terraform v1.7.5
on linux_amd64
+ provider registry.terraform.io/ciscodevnet/sdwan v0.3.8

$ terraform apply
...
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the
following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # sdwan_switchport_feature_template.DELETE_ME will be updated in-place
  ~ resource "sdwan_switchport_feature_template" "DELETE_ME" {
        id            = "8c64b99f-6c9e-4aee-b62f-e563be447e2f"
      ~ interfaces    = [
          ~ {
              + dot1x_enable                    = false
                name                            = "GigabitEthernet0/1/0"
                # (3 unchanged attributes hidden)
            },
          ~ {
              + dot1x_enable                    = true
                name                            = "GigabitEthernet0/1/1"
                # (3 unchanged attributes hidden)
            },
            # (1 unchanged element hidden)
        ]
        name          = "DELETE_ME"
      ~ version       = 0 -> (known after apply)
        # (6 unchanged attributes hidden)
    }
Plan: 0 to add, 1 to change, 0 to destroy.
....
sdwan_switchport_feature_template.DELETE_ME: Modifying... [id=8c64b99f-6c9e-4aee-b62f-e563be447e2f]
sdwan_switchport_feature_template.DELETE_ME: Modifications complete after 2s [id=8c64b99f-6c9e-4aee-b62f-e563be447e2f]

```
4) rinse and repeat, next "apply" action will be "Modifying" without doing actual change

**Comment by danischm** [comment]
Could you share your full resource config? Do you have other dot1x attributes configured? This works for me:

```terraform
resource "sdwan_switchport_feature_template" "example" {
  name         = "Example"
  description  = "My Example"
  device_types = ["vedge-C8000V"]
  interfaces = [
    {
      name         = "GigabitEthernet0/0/0"
      dot1x_enable = null
    }
  ]
}
```

**Comment by danischm** [comment]
I see what you mean.. if you edit it, it always shows as enabled.

**Comment by cmohorea** [comment]
its not a matter of "show": if you edit, it won't be set on the vManage (e.g. go to GUI and edit). 
But, if you set it to "false" and then manually update on vManage to "Off", then TF will be in sync and won't complain

**Comment by danischm** [comment]
This should fix an issue where certain elements (like `dot1x_enable`) where omitted from the payload: https://github.com/CiscoDevNet/terraform-provider-sdwan/commit/f56641490b33e75569472eb3f2996c0f092f7682

Not sure if you are familiar with GO, but if you want to try this out, you can install the provider binary locally (using `go install`) and then use a TF development override (https://developer.hashicorp.com/terraform/cli/config/config-file#development-overrides-for-provider-developers) to refer to your local binary.

**Comment by danischm** [comment]
Published in v0.3.9 release.
